File: src/client/java/net/greenjab/fixedminecraft/map_book/MapTile.java
package net.greenjab.fixedminecraft.map_book;

import net.greenjab.fixedminecraft.screens.MapBookScreen;
import net.minecraft.client.Minecraft;
import net.minecraft.client.gui.GuiGraphicsExtractor;
import net.minecraft.client.gui.components.Renderable;
import net.minecraft.client.renderer.state.MapRenderState;
import net.minecraft.world.level.saveddata.maps.MapId;
import net.minecraft.world.level.saveddata.maps.MapItemSavedData;
import org.joml.Matrix3x2fStack;

/** Credit: Nettakrim */
public class MapTile implements Renderable {
    private final MapBookScreen screen;
    MapId id;
    private final MapItemSavedData mapState;

    private final MapRenderState mapRenderState = new MapRenderState();

    public MapTile(MapBookScreen screen, MapId id, MapItemSavedData mapState, Minecraft client) {
        this.screen = screen;
        this.id = id;
        this.mapState = mapState;
        client.getMapRenderer().extractRenderState(id, mapState, mapRenderState);
    }

    @Override
    public void extractRenderState(GuiGraphicsExtractor context, int mouseX, int mouseY, float delta) {
        float mapScale = (float) Math.pow(2, mapState.scale);
        float offset = 64f * mapScale;

        Matrix3x2fStack matrix = context.pose();

        matrix.pushMatrix();
        //for (int i = 0; i < 4-mapState.scale;i++)context.goUpLayer();
        matrix.translate(screen.x, screen.y);
        matrix.scale(screen.scale, screen.scale);
        matrix.translate(
                (float) (mapState.centerX - offset + screen.width / 2.0),
                (float) (mapState.centerZ - offset + screen.height / 2.0));
        matrix.scale(mapScale, mapScale);
        context.map(mapRenderState);
        //for (int i = 0; i < 4-mapState.scale;i++)context.popLayer();
        matrix.popMatrix();
    }
}

File: src/client/kotlin/net/greenjab/fixedminecraft/map_book/MapTile.kt
package net.greenjab.fixedminecraft.map_book

import net.greenjab.fixedminecraft.mixin.client.map.DrawContextAccessor
import net.minecraft.client.MinecraftClient
import net.minecraft.client.gui.DrawContext
import net.minecraft.client.gui.Drawable
import net.minecraft.client.render.MapRenderState
import net.minecraft.component.type.MapIdComponent
import net.minecraft.item.map.MapState


class MapTile(private val screen: MapBookScreen, id: MapIdComponent?, private val mapState: MapState, private val client: MinecraftClient) :
    Drawable {
    private val mapRenderState = MapRenderState()

    init {
        client.mapRenderer.update(id, mapState, mapRenderState)
    }

    override fun render(context: DrawContext, mouseX: Int, mouseY: Int, delta: Float) {
        val mapScale = (1 shl mapState.scale.toInt()).toFloat()
        val offset = 64f * mapScale
        context.matrices.push()
        context.matrices.translate(screen.x, screen.y, 1.0 / (mapState.scale.toDouble() + 1.0) + 1.0)
        context.matrices.scale(screen.scale, screen.scale, -1.0f)
        context.matrices.translate(
            mapState.centerX.toDouble() - offset.toDouble() + screen.width.toDouble() / 2.0,
            mapState.centerZ.toDouble() - offset.toDouble() + screen.height.toDouble() / 2.0, 0.0
        )
        context.matrices.scale(mapScale, mapScale, 1.0f)
        client.mapRenderer.draw(mapRenderState, context.matrices, (context as DrawContextAccessor).vertexConsumers, true, 15728880)
        context.matrices.pop()
    }
}
